File: Python/Clase6/Clase6_Teoria.py
## Clase6_Python_Teoria
##Programación Orientada a Objetos  (POO)

#Creación de una clase

class Persona: #Creamos una clase
   
    # init es un MÉTODO, lo que está dentro son sus atributos,
    def __init__(self, nombre, apellido, edad): #Self se asigna por default y es la referencia al objeto. #Luego se le pasan los parámetros (ver lin 9-11)
      self.nombre = nombre         #Creamos una variable nombre que se pasará como parámetro. Antes, valor hardcodeado: 'Juan'
      self.apellido = apellido       #Creamos una variable apellido que se pasará como parámetro.  Antes, valor harcodeado: 'Perez'
      self.edad = edad         #Creamos una variable edad que se pasará como parámetro. Antes, valor harcodeado: 22
    def mostrar_detalle(self):
       print(f'Persona: {self.nombre} {self.apellido} {self.edad}')


print(Persona)

##Objeto persona1 
persona1 = Persona('Nicolas', 'Giglio', 37) #No pasamos ninguna referencia. Es una instancia que referencia al constructor (init) de manera indirecta.
print(persona1.nombre)
print(persona1.apellido)
print(persona1.edad)
print(f'El objeto 1 de la clase Persona: {persona1.nombre}, {persona1.apellido}, y su edad es: {persona1.edad}') #Tarea: imprimir en una linea.

##Objeto persona2
persona2 = Persona('Federico', 'Giglio', 33)
print(f'El objeto 2 de la clase Persona: {persona2.nombre}, {persona2.apellido}, y su edad es {persona2.edad}')


#Modificando los atributos de un objeto

persona1.nombre = 'Tito'
persona1.apellido = 'Puente'
persona1.edad = 70

print(f'El objeto 1 tiene ahora los siguientes atributos: nombre: {persona1.nombre}, apellido: {persona1.apellido}, edad: {persona1.edad}')

#Recordar:
## Los atributos son las CARACTERÍSTICAS que va a tener un objeto.
## Los métodos son las ACCIONES que van a ejecutar los objetos. (comportamiento)

persona1.mostrar_detalle() #Llamamos al método "mostrar_detalle" dentro del objeto (creado en lin 13)
persona2.mostrar_detalle()

File: JavaScript/Clase6/Clase6_Teoria.js
//Clase6_JavaScript_Teoría

//Clases
//La clase es una plantilla, en la cual definimos los atributos/metodos que van a tener los objetos (instancias) que contenga.

//Sintáxis básica para crear una clase
class Persona{ //Clase Padre (Parent)
    constructor(nombre, apellido){      //Si no se define un constructor, JS crea uno de manera automática.
        this._nombre = nombre;
        this._apellido = apellido;  //Buena práctica: los atributos de la clase no pueden llamarse igual cuando las llamamos (get/set), por eso se usa un guion.

    };

    get nombre(){
        return this._nombre;
    };
    
    set nombre(nombre){
        return this._nombre = nombre;
    };

    get apellido(){
        return this._apellido;
    };

    set apellido(apellido){
        return this._apellido = apellido;
    };

}

//Herencia
class Empleado extends Persona{ //Clase Hija (Child) --> Ver Diagrama UML
    constructor (nombre, apellido, departamento){
        super(nombre, apellido) //Tenemos que llamar al constructor de la CLASE PADRE usando super() EN LA PRIMERA LÍNEA del constructor Child.
        this._departamento = departamento;
    }

    get departamento(){
        return this._departamento;
    }

    set departamento(departamento){
        this._departamento = departamento;
    }
}

let persona1 = new Persona('Martin', 'Perez');
//console.log(persona1);
console.log(persona1.nombre); //llamamos al atributo con el método get (está dentro de los paréntesis ".") (definido en la clase)
persona1.nombre = 'Juan Carlos'; //cambiamos el nombre usando el método set (definido en la clase)
console.log(persona1.nombre);
console.log(persona1.apellido); //Traemos el atributo actual del objeto (definido en lin 18)
persona1.apellido = 'Martinez'; //Modificamos el segundo atributo con el método get.
console.log(persona1); //Imprmimios el objeto con los nuevos atributos
let persona2 = new Persona('Carlos', 'Lara');
//console.log(persona2);
console.log(persona2.nombre);
persona2.nombre = 'María Laura';
console.log(persona2.apellido);
persona2.apellido = 'Castro';
console.log(persona2);

//Hoisting 
//NO funciona para clases, es decir, no puedo llamar una instancia de la clase antes de inicializarla/definirla.

//HERENCIA
//Creamos un objeto de la clase Child (Definida en línea 33)
let empleado1 = new Empleado('Nicolas', 'Giglio', 'QA');
console.log(empleado1);
console.log(empleado1._nombre) //El atributo "nombre" está HEREDADO de la clase PARENT (Persona).
